packages/firestore/src/local/indexeddb_persistence.ts

```diff
@@ -64,7 +64,7 @@ const CLIENT_METADATA_MAX_AGE_MS = 5000;
  * garbage collected. Clients that exceed this threshold will not be able to
  * replay Watch events that occurred before this threshold.
  */
-const CLIENT_METADATA_GARBAGE_COLLECTION_THRESHOLD_MS = 30 * 60 * 1000; // 30 Minutes
+const CLIENT_STATE_GARBAGE_COLLECTION_THRESHOLD_MS = 30 * 60 * 1000; // 30 Minutes
```

Comment:
[adding this comment here out of laziness]

Trying to reconcile the two timeouts here (CLIENT_METADATA_MAX_AGE_MS and CLIENT_STATE_GARBAGE_COLLECTION_THRESHOLD_MS), I think:

1. CLIENT_METADATA_MAX_AGE_MS is basically how stale a client can be and still be eligible to be primary.
2. CLIENT_STATE_GARBAGE_COLLECTION_THRESHOLD_MS is basically how stale a client can be and still "catch up" to current.

If that's right, we might want to do some renaming / comment tweaks...  But I also think the fact that `getActiveClients()` uses CLIENT_METADATA_MAX_AGE_MS is kinda' wrong.  If a primary change happens, any clients more stale than CLIENT_METADATA_MAX_AGE_MS will be ignored by the new primary and we won't start listeners for them.  So they will not properly "catch up" until they perform some extra action to re-persist their client state to local storage so that the primary notices them.

On that note, I'm wondering if we might want to ditch `getActiveClients()` and instead have the primary client write some sort of "report in please" notification to LocalStorage that triggers all secondary tabs to (re-)write their current state.  We could then use that mechanism on primary changes in order to get existing clients, rather than using `getActiveClients()`.  I think that might be cleaner and easier to reason about (we don't have an "active client" notion that crosses over between IndexedDbPersistence and SharedClientState)... WDYT?

Reply:
I don't think I want to re-work the source of how we determine what clients are active. IndexedDb seems to be the more accurate source of truth here, and it has the only timestamp that actually gets updated (note that the row in LocalStorage has an updateTime, and I am wondering if I should remove it since it does not get refreshed).


As for everything else you said, I think you are correct: CLIENT_METADATA_MAX_AGE_MS is (right now) the age at which clients are considered active, and CLIENT_STATE_GARBAGE_COLLECTION_THRESHOLD_MS is when we "give up on the client". I think you make a good case that we should not consider clients inactive that we don't yet "give up" on. It might make sense to use the 30 minute interval everywhere, and only use the 5s interval in the primary tab election. Since it is Friday, I am going to submit this PR, but I did file b/112478723.